src/ca/verify.ts

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2022 The Sigstore Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import { VerificationError } from '../error';
+import * as sigstore from '../types/sigstore';
+import { x509Certificate } from '../x509/cert';
+import { verifyCertificateChain } from '../x509/verify';
+
+export function verifySigningCertificate(
+  bundle: sigstore.Bundle,
+  trustedRoot: sigstore.TrustedRoot,
+  options: sigstore.ArtifactVerificationOptions_CtlogOptions
+) {
+  if (bundle.verificationMaterial?.content?.$case !== 'x509CertificateChain') {
+    throw new VerificationError('No certificate chain in bundle');
+  }
+
+  const bundleCerts = parseCerts(
+    bundle.verificationMaterial.content.x509CertificateChain.certificates
+  );
+
+  // Check that a trusted certificate chain can be found for the signing
+  // certificate in the bundle
+  const trustedChain = verifyChain(
+    bundleCerts,
+    trustedRoot.certificateAuthorities
+  );
+
+  // Unless disabled, verify the SCTs in the signing certificate
+  if (!options.disable) {
+    verifySCTs(trustedChain, trustedRoot.ctlogs, options);
+  }
+}
+
+function verifyChain(
+  bundleCerts: x509Certificate[],
+  certificateAuthorities: sigstore.CertificateAuthority[]
+): x509Certificate[] {
+  const signingCert = bundleCerts[0];
+
+  // Filter the list of certificate authorities to those which are valid for the
+  // signing certificate's notBefore date.
+  const validCAs = filterCertificateAuthorities(
+    certificateAuthorities,
+    signingCert.notBefore
+  );
+
+  if (validCAs.length === 0) {
+    throw new VerificationError('No valid certificate authorities');
+  }
+
+  let trustedChain: x509Certificate[] = [];
+
+  // Loop through all valid CAs and attempt to verify the certificate chain
+  const verified = validCAs.find((ca) => {
```

Comment:
Does `find` not return as soon as it you return true? Should this be a map?

Reply:
or wait, you only care that one valid CA has a valid cert chain? the comment tripped me up